src/Xrefcheck/Verify.hs

```diff
@@ -81,230 +87,324 @@ toVerifyRes = VerifyResult . either one (\() -> [])
+        & maybe (throwError ExternalResourceUnknownProtocol) pure
+      let reqLink = case parsedUrl of
+            Left (url, option) ->
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+            Right (url, option) ->
+              runReq defaultHttpConfig $
+              req method url NoReqBody ignoreResponse option
+
+      let maxTime = Time @Second $
+            unTime vcExternalRefCheckTimeout * timeoutFrac
+
+      mres <- liftIO (timeout maxTime $ void reqLink) `catch`
+        (either throwError (\() -> return (Just ())) . interpretErrors)
+      maybe (throwError $ ExternalResourceSomeError "Response timeout") pure mres
 
     isAllowedErrorCode = or . sequence
-        -- We have to stay conservative - if some URL can be accessed under
-        -- some circumstances, we should do our best to report it as fine.
-        [ (403 ==)  -- unauthorized access
-        , (405 ==)  -- method mismatch
-        ]
+      -- We have to stay conservative - if some URL can be accessed under
+      -- some circumstances, we should do our best to report it as fine.
+      [ (403 ==)  -- unauthorized access
+      , (405 ==)  -- method mismatch
+      ]
 
     interpretErrors = \case
-        JsonHttpException _ -> error "External link JSON parse exception"
-        VanillaHttpException err -> case err of
-            InvalidUrlException{} -> error "External link URL invalid exception"
-            HttpExceptionRequest _ exc -> case exc of
-                StatusCodeException resp _
-                    | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
-                    | otherwise -> Left $ ExternalResourceUnavailable (responseStatus resp)
-                other -> Left . ExternalResourceSomeError $ show other
+      JsonHttpException _ -> error "External link JSON parse exception"
+      VanillaHttpException err -> case err of
+        InvalidUrlException{} -> error "External link URL invalid exception"
+        HttpExceptionRequest _ exc -> case exc of
+          StatusCodeException resp _
+            | isAllowedErrorCode (statusCode $ responseStatus resp) -> Right ()
+            | otherwise -> Left $
+                ExternalHttpResourceUnavailable (responseStatus resp)
+          other -> Left . ExternalResourceSomeError $ show other
+
+    checkFtp :: URI -> Bool -> ExceptT VerifyError IO ()
+    checkFtp uri secure = do
+      -- get authority which stores host and port
+      authority <- case uriAuthority uri of
+        Right a -> pure a
+        Left _ -> throwError ExternalResourceInvalidUri
+      let host = T.unpack . unRText $ authHost authority
+      let port :: Int = fromIntegral . fromMaybe 21 $ authPort authority
+      let pieces = uriPath uri
+      -- in case there aro no pieces - smth wrong with url
+      unless (isJust pieces) $
+        throwError ExternalResourceInvalidUri
+      let (trailing, pathList) = fromJust pieces
```

Comment:
We shouldn't use `fromJust`.
We should pattern match on `pieces :: Maybe etc`, and then branch out.

Yes, in this instance, it technically _is_ safe to use `fromJust` here, because we're guarded by `isJust` 2 lines up.
_But_, during a refactoring, while moving things around, it's easy to move/copy-paste the `fromJust` somewhere else, and leave the `isJust` behind. When that happens, the code will still compile, but then fail to run.
I know it sounds a bit farfetched, but this does happen in practice.

For this reason, pattern matching and branching out is considered a more "refactoring-safe" solution.

Reply:
I've got it! thanks